hopsworks-api/src/main/java/io/hops/hopsworks/api/provenance/ProvenanceCleaner.java

```diff
@@ -0,0 +1,143 @@
+
+import io.hops.hopsworks.common.dao.hdfs.inode.Inode;
+import io.hops.hopsworks.common.dao.hdfs.inode.InodeFacade;
+import io.hops.hopsworks.common.dao.project.Project;
+import io.hops.hopsworks.common.dao.project.ProjectFacade;
+import io.hops.hopsworks.common.provenance.core.elastic.ProvElasticController;
+import io.hops.hopsworks.common.util.Settings;
+import io.hops.hopsworks.exceptions.ElasticException;
+import io.hops.hopsworks.exceptions.ProvenanceException;
+import io.hops.hopsworks.restutils.RESTCodes;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.admin.indices.delete.DeleteIndexRequest;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.indices.GetIndexRequest;
+import org.elasticsearch.client.indices.GetIndexResponse;
+import org.elasticsearch.rest.RestStatus;
+import org.javatuples.Pair;
+
+import javax.ejb.EJB;
+import javax.ejb.Schedule;
+import javax.ejb.Singleton;
+import javax.ejb.Timer;
+import javax.ejb.TransactionAttribute;
+import javax.ejb.TransactionAttributeType;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Singleton
+@TransactionAttribute(TransactionAttributeType.NOT_SUPPORTED)
+public class ProvenanceCleaner {
+  private final static Logger LOGGER = Logger.getLogger(ProvenanceCleaner.class.getName());
+  
+  @EJB
+  private Settings settings;
+  @EJB
+  private ProvElasticController client;
+  @EJB
+  private ProjectFacade projectFacade;
+  @EJB
+  private InodeFacade inodeFacade;
+  
+  private String lastIndexChecked = "";
+  
+  // Run once per hour
+  @Schedule(persistent = false, hour = "*", minute = "*")
+  public void execute(Timer timer) {
+    int cleanupSize = settings.getProvCleanupSize();
+    int archiveSize = settings.getProvArchiveSize();
+    if(archiveSize == 0) {
+      return;
+    }
+    try {
+      Pair<Integer, String> round = archiveRound(lastIndexChecked, cleanupSize);
+      LOGGER.log(Level.INFO, "cleanup round - idx cleaned:{0} from:{1} to:{2}",
+        new Object[]{round.getValue0(), lastIndexChecked, round.getValue1()});
+      lastIndexChecked = round.getValue1();
+    } catch (ProvenanceException | ElasticException e) {
+      LOGGER.log(Level.INFO, "cleanup round was not successful - error", e);
+    }
+  }
+  
+  private Pair<Integer, String> archiveRound(String nextToCheck, Integer limitIdx)
+    throws ProvenanceException, ElasticException {
+    String[] indices = getAllIndices();
+    
+    int cleaned = 0;
+    String nextToCheckAux = "";
+    for(String indexName : indices) {
+      if(cleaned > limitIdx) {
+        nextToCheckAux = indexName;
+        break;
+      }
+      if(indexName.compareTo(nextToCheck) < 0) {
+        continue;
+      }
+      Project project = getProject(indexName);
+      if(project == null) {
+        LOGGER.log(Level.INFO, "deleting prov index:{0} with no corresponding project", indexName);
+        deleteProvIndex(indexName);
```

Comment:
Here you are not catching the exception coming from `deleteProvIndex` if for some reason elastic is not able to delete the first index - you sill be always stuck in trying to delete the same index.

Reply:
So the exception will get thrown and interrupt the for loop in this batch. Yes on next batch you will try again. You will be stuck here, but if it is a transient error (elastic is down) so eventually it will get fixed, or it is a logic error/internal error issue, which needs manual intervention anyway. I don't have a dead letter box type of abstraction.